com.unity.multiplayer.mlapi/Runtime/AOI/ClientObjectMap.cs

```diff
@@ -0,0 +1,110 @@
+// Although this is (currently) inside the MLAPI package, it is intentionally
+//  totally decoupled from MLAPI with the intention of allowing it to live
+//  in its own package
+namespace MLAPI.AOI
+{
+    // To establish a Client Object Map, instantiate a ClientObjMapNodeBase, then
+    //  add more nodes to it (and those nodes) as desired
+   public class ClientObjMapNode<CLIENT, OBJECT>
+   {
+        // set this delegate if you want a function called when
+        //  object 'obj' is being de-spawned
+        public delegate void DespawnDelegate(OBJECT obj);
+        public DespawnDelegate OnDespawn;
+
+        // to dynamically compute objects to be added each 'QueryFor' call,
+        //  assign this delegate to your handler
+        public delegate void QueryDelegate(CLIENT c, HashSet<OBJECT> results);
+        public QueryDelegate OnQuery;
+
+        public delegate void BypassDelegate(HashSet<OBJECT> results);
+        public BypassDelegate OnBypass;
+
+        public ClientObjMapNode() : base()
+        {
+            ChildNodes = new List<ClientObjMapNode<CLIENT, OBJECT>>();
+        }
+
+        // externally-called object query function.  Call this on your root
+        //  ClientObjectMapNode.  The passed-in hash set will contain the results.
+        public void QueryFor(CLIENT client, HashSet<OBJECT> results)
+        {
+            if (Bypass)
+            {
+                OnBypass(results);
+            }
+            else
+            {
+                if (OnQuery != null)
+                {
+                    OnQuery(client, results);
+                }
+
+                foreach (var c in ChildNodes)
+                {
+                    c.QueryFor(client, results);
+                }
+            }
+        }
+
+        // Called when a given object is about to be despawned.  The OnDespawn
+        //  delegate gives each node a chance to do its own handling (e.g. removing
+        //  the object from a cache)
+        public void DespawnCleanup(OBJECT o)
+        {
+            if (OnDespawn != null)
+            {
+                OnDespawn(o);
+            }
+
+            foreach (var c in ChildNodes)
+            {
+                c.DespawnCleanup(o);
+            }
+        }
+
+        // Add a new child node.  Currently, there is no way to remove a node
+        public void AddNode(ClientObjMapNode<CLIENT, OBJECT> @new)
+        {
+            ChildNodes.Add(@new);
+        }
+
+        private List<ClientObjMapNode<CLIENT, OBJECT>> ChildNodes;
```

Comment:
I think this could be named as `_childNodes` (underscore and lowercase first letter for private members)